src/plugin/adapter-registry/amalgamator-gdb-tracker.ts

```diff
@@ -0,0 +1,114 @@
+/********************************************************************************
+ * Copyright (C) 2024 Ericsson, Arm and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { DebugProtocol } from '@vscode/debugprotocol';
+import * as vscode from 'vscode';
+import { Context, ReadMemoryArguments, ReadMemoryResult, WriteMemoryArguments, WriteMemoryResult } from '../../common/messaging';
+import { AdapterCapabilities, AdapterVariableTracker, VariableTracker } from './adapter-capabilities';
+
+// Copied from cdt-amalgamator [AmalgamatorSession.d.ts] file
+/**
+ * Response for our custom 'cdt-amalgamator/getChildDaps' request.
+ */
+export interface Contexts {
+    children?: Context[];
+}
+export interface GetContextsResponse extends DebugProtocol.Response {
+    body: Contexts;
+}
+export type GetContextsResult = GetContextsResponse['body'];
+
+export interface AmalgamatorReadArgs extends ReadMemoryArguments {
+    child: Context;
+}
+
+export class AmalgamatorSessionManager extends VariableTracker implements AdapterCapabilities {
```

Comment:
It sounds like we likely want to move this over to the Amalgamator repository and make it dependent on this plugin (or at least propose it).

Reply:
I agree with @colin-grant-work . I think this is a great example of how a custom adapter implementation could look like.
But ideally it would be contributed by the debug adapter. Or was the intention to use it in a downstream version of the Memory Inspector, @WyoTwT ?

Maybe we can build some "how to implement your own adapter contribution" docs around it in https://github.com/eclipse-cdt-cloud/vscode-memory-inspector?tab=readme-ov-file#the-memory-provider ? It could be a nice story to demonstrate this with two Eclipse CDT Cloud components.

Another thing this PR shows me is that we may want to consider to somehow provide access to the default adapter implementation for reuse in custom implementations. Beyond asking developers to include the according sources from this repo.

Reply:
It doesn't appear there is an easy way to get access to the AdapterCapabilities, AdapterVariableTracker, and VariableTracker if this code is moved to the Amalgamator repository. 

I could duplicate the interfaces in Amalgamator which is what I did with the Context because that wasn't shared from the Amalgamator.  Would restructuring the code differently facilitate future movement of some parts into a form that could be reused for custom implementations?

I'm rather new to this so all suggestions to better structure the code are welcome but I may need additional clarification.